### PR TITLE
[cudapoa] improving cudaPOA performance

### PR DESCRIPTION
**Pull request by r-mafi**
a round of optimization consisting of reducing back register usage, hiding global memory access where applicable and reducing NW inner while loop iterations in an effort to improve compute-time and SOL metrics.
in cudapoa binary API, added a new option `-s` to allow managing allocated memory for adaptive score matrix

### CODE REVIEW

File: cudapoa/src/cudapoa_kernels.cuh [hunk 1]
```diff
@@ -28,6 +28,8 @@
 #include <claraparabricks/genomeworks/utils/cudautils.hpp>
 #include <claraparabricks/genomeworks/cudapoa/batch.hpp>
 
+#define GW_POA_KERNELS_MAX_THREADS_PER_BLOCK 896
```
[the remark this answers is not in the document]
Comment (tijyojwad) [reply cudapoa/src/cudapoa_kernels.cuh:31]: can you add a comment for this?

File: cudapoa/src/cudapoa_nw_banded.cuh [hunk 2]
```diff
@@ -297,46 +299,39 @@ __device__
 
             while (__any_sync(FULL_MASK, loop))
             {
-
-                // To increase instruction level parallelism, we compute the scores
```
Comment (tijyojwad) [inline cudapoa/src/cudapoa_nw_banded.cuh:301]: and can you add the comment here that this depends on the profile of the reads - more accurate/short read data may benefit from the ILP vs the less accurate read data benefits from the reduced while loop count?

**Review by tijyojwad — CHANGES_REQUESTED** [review]

File: cudapoa/src/cudapoa_kernels.cuh (hunk 1 above)
[the remark this answers is not in the document]
Comment (r-mafi) [reply cudapoa/src/cudapoa_kernels.cuh:31]: with some more changes, I was able to bring down register count in most cases down to 64. So changed 896 to 1024, and added a comment, explaining with 65536 register file size, __launch_bounds__(1024) translates to max 64 register usage.

File: cudapoa/src/cudapoa_nw_banded.cuh (hunk 2 above)
Comment (r-mafi) [reply cudapoa/src/cudapoa_nw_banded.cuh:301]: done

**Review by tijyojwad — APPROVED** [review]

**Pull request merged**